File: backend/app/services/auth_service.py
"""
Authentication service matching Java AuthService.java.
Handles registration and login for all user roles.
"""
from sqlalchemy.orm import Session

from app.core.security import hash_password, verify_password, create_access_token, get_expiration_seconds
from app.exceptions.custom_exceptions import DuplicateResourceException, ResourceNotFoundException, BadCredentialsException, DisabledException
from app.models.user import User, Role
from app.models.candidate import Candidate
from app.models.recruiter import Recruiter
from app.repositories.user_repository import UserRepository
from app.repositories.candidate_repository import CandidateRepository
from app.repositories.recruiter_repository import RecruiterRepository
from app.repositories.company_repository import CompanyRepository
from app.schemas.auth import (
    RegisterCandidateRequest,
    RegisterRecruiterRequest,
    LoginRequest,
    LoginResponse,
    UserResponse
)


class AuthService:
    """
    Handles the three public entry points into the system: candidate
    registration, recruiter registration, and login.
    Matches Java AuthService.java.
    """
    
    def __init__(self, db: Session):
        self.db = db
        self.user_repo = UserRepository(db)
        self.candidate_repo = CandidateRepository(db)
        self.recruiter_repo = RecruiterRepository(db)
        self.company_repo = CompanyRepository(db)
    
    def register_candidate(self, request: RegisterCandidateRequest) -> UserResponse:
        """
        Register a new candidate user.
        Matches Java: @Transactional public UserResponse registerCandidate(...)
        """
        email = request.email.lower().strip()
        
        if self.user_repo.exists_by_email(email):
            raise DuplicateResourceException("An account with this email already exists.")
        
        # Create user
        user = User(
            name=request.name.strip(),
            email=email,
            password_hash=hash_password(request.password),
            role=Role.CANDIDATE,
            enabled=True
        )
        user = self.user_repo.save(user)
        
        # Create candidate profile
        candidate = Candidate(id=user.id)
        self.db.add(candidate)
        self.db.commit()
        
        return self._user_to_response(user)
    
    def register_recruiter(self, request: RegisterRecruiterRequest) -> UserResponse:
        """
        Register a new recruiter user.
        Matches Java: @Transactional public UserResponse registerRecruiter(...)
        """
        email = request.email.lower().strip()
        
        if self.user_repo.exists_by_email(email):
            raise DuplicateResourceException("An account with this email already exists.")
        
        # Find company if specified
        company = None
        if request.company_name and request.company_name.strip():
            company = self.company_repo.find_by_name_ignore_case(request.company_name.strip())
            if not company:
                raise ResourceNotFoundException(
                    f"No company found named '{request.company_name}'. "
                    "Ask an admin to create it first, or register without a company for now."
                )
        
        # Create user
        user = User(
            name=request.name.strip(),
            email=email,
            password_hash=hash_password(request.password),
            role=Role.RECRUITER,
            enabled=True
        )
        user = self.user_repo.save(user)
        
        # Create recruiter profile
        recruiter = Recruiter(
            id=user.id,
            designation=request.designation,
            company_id=company.id if company else None
        )
        self.db.add(recruiter)
        self.db.commit()
        
        return self._user_to_response(user)
    
    def login(self, request: LoginRequest) -> LoginResponse:
        """
        Authenticate user and generate JWT token.
        Matches Java: public LoginResponse login(LoginRequest request)
        """
        email = request.email.lower().strip()
        
        user = self.user_repo.find_by_email(email)
        if not user:
            raise BadCredentialsException("Invalid email or password")
        
        if not verify_password(request.password, user.password_hash):
            raise BadCredentialsException("Invalid email or password")
        
        if not user.enabled:
            raise DisabledException("This account has been disabled")
        
        # Generate JWT token
        token = create_access_token(user.email, user.role.value)
        
        return LoginResponse(
            token=token,
            token_type="Bearer",
            expires_in_seconds=get_expiration_seconds(),
            user=self._user_to_response(user)
        )
    
    def _user_to_response(self, user: User) -> UserResponse:
        """Convert User entity to UserResponse DTO."""
        return UserResponse(
            id=user.id,
            name=user.name,
            email=user.email,
            role=user.role,
            enabled=user.enabled
        )

File: backend/src/main/java/com/codeforge/service/AuthService.java
package com.codeforge.service;

import com.codeforge.dto.request.LoginRequest;
import com.codeforge.dto.request.RegisterCandidateRequest;
import com.codeforge.dto.request.RegisterRecruiterRequest;
import com.codeforge.dto.response.LoginResponse;
import com.codeforge.dto.response.UserResponse;
import com.codeforge.entity.Candidate;
import com.codeforge.entity.Recruiter;
import com.codeforge.entity.User;
import com.codeforge.exception.BadCredentialsException;
import com.codeforge.exception.DuplicateResourceException;
import com.codeforge.repository.CandidateRepository;
import com.codeforge.repository.RecruiterRepository;
import com.codeforge.repository.UserRepository;
import com.codeforge.security.JWTProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Authentication service for user registration and login
 */
@Service
@RequiredArgsConstructor
public class AuthService {
    
    private final UserRepository userRepository;
    private final CandidateRepository candidateRepository;
    private final RecruiterRepository recruiterRepository;
    private final JWTProvider jwtProvider;
    private final PasswordEncoder passwordEncoder;
    
    /**
     * Register a new candidate
     */
    @Transactional
    public UserResponse registerCandidate(RegisterCandidateRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw DuplicateResourceException.alreadyExists("User", "email", request.getEmail());
        }
        
        User user = User.builder()
            .name(request.getName())
            .email(request.getEmail())
            .passwordHash(passwordEncoder.encode(request.getPassword()))
            .role(User.UserRole.CANDIDATE)
            .enabled(true)
            .build();
        
        user = userRepository.save(user);
        
        // Create candidate profile
        Candidate candidate = Candidate.builder()
            .id(user.getId())
            .user(user)
            .build();
        candidateRepository.save(candidate);
        
        return mapToUserResponse(user);
    }
    
    /**
     * Register a new recruiter
     */
    @Transactional
    public UserResponse registerRecruiter(RegisterRecruiterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw DuplicateResourceException.alreadyExists("User", "email", request.getEmail());
        }
        
        User user = User.builder()
            .name(request.getName())
            .email(request.getEmail())
            .passwordHash(passwordEncoder.encode(request.getPassword()))
            .role(User.UserRole.RECRUITER)
            .enabled(true)
            .build();
        
        user = userRepository.save(user);
        
        // Create recruiter profile
        Recruiter recruiter = Recruiter.builder()
            .id(user.getId())
            .user(user)
            .phone(request.getPhone())
            .position(request.getPosition())
            .build();
        recruiterRepository.save(recruiter);
        
        return mapToUserResponse(user);
    }
    
    /**
     * Login user and return JWT token
     */
    public LoginResponse login(LoginRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
            .orElseThrow(() -> new BadCredentialsException("Invalid email or password"));
        
        if (!user.getEnabled()) {
            throw new BadCredentialsException("User account is disabled");
        }
        
        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new BadCredentialsException("Invalid email or password");
        }
        
        String token = jwtProvider.generateToken(user.getId(), user.getEmail(), user.getRole().name());
        
        return LoginResponse.builder()
            .token(token)
            .tokenType("Bearer")
            .expiresInSeconds(86400) // 24 hours
            .user(mapToUserResponse(user))
            .build();
    }
    
    private UserResponse mapToUserResponse(User user) {
        return UserResponse.builder()
            .id(user.getId())
            .name(user.getName())
            .email(user.getEmail())
            .role(user.getRole().name())
            .enabled(user.getEnabled())
            .build();
    }
}
